### PR TITLE
Fix matlab paths

### PR DESCRIPTION
**Pull request by klshrinidhi**
This is to fix configure script for Matlab support.
See #660, #44.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -377,6 +377,19 @@ if(${BUILD_JAVA_WRAPPING})
...
+        if(${Matlab_MEX_EXTENSION} MATCHES "64")
+            # On Unix, assume host architecture to be same as target 
+            # architecture.
+            if((UNIX AND NOT ${CMAKE_HOST_SYSTEM_PROCESSOR} MATCHES "64") OR
+               (MSVC AND NOT CMAKE_GENERATOR MATCHES "64"))
+                message(WARNING 
+                    "Matlab found is 64-bit but C++ compiler is 32-bit.")
+            endif()
+        endif()
+    endif()
```
Comment (chrisdembia) [inline CMakeLists.txt]: I think if a user has 32-bit MATLAB but a 64-bit compiler, they will not get a warning.

Comment (chrisdembia) [inline CMakeLists.txt]: I see that you assume on UNIX that host and target are the same, but the message could take this into account better. Perhaps give a different warning on UNIX: "Matlab found is 64-bit but host architecture is 32-bit."

File: Bindings/Java/Matlab/configureOpenSim.m.in [hunk 2]
```diff
@@ -49,18 +46,36 @@
...
     end
 %% Edit the classpath.txt file (need full path for print)
-% Create the string names used
+    % Create the string names used
     classFile = fullfile(mr, 'toolbox', 'local', 'classpath.txt');
-    OpenSimJarPath =  fullfile(openSimFolder, 'opensim','modules','org-opensim-modeling.jar');
-% Edit the class path txt file
-    edit_path_txt_file(classFile,OpenSimJarPath)
-
+    OpenSimJarPath =  fullfile(openSimFolder, 'share', 'java', 'org-opensim-modeling.jar');
```
Comment (chrisdembia) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: This is not cross-platform. Must configure from OPENSIM_INSTALL_JAVAJARDIR

File: Bindings/Java/Matlab/configureOpenSim.m.in [hunk 3]
```diff
@@ -49,18 +46,36 @@
...
+    % Edit the class path txt file
+    success = edit_path_txt_file(classFile, OpenSimJarPath);
+    % If above failed, Matlab is probably new. Try newer technique.
+    if ~success
+        classFile = fullfile(prefdir, 'javaclasspath.txt');
+        success = edit_path_txt_file(classFile, OpenSimJarPath);
+        if ~success
+            h = msgbox('Failed to add classpath.', 'Error');
```
Comment (chrisdembia) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: "Failed to add _to_ class path"

File: Bindings/Java/Matlab/configureOpenSim.m.in [hunk 4]
```diff
@@ -112,34 +135,32 @@ end
...
+        fclose(fileID);
+        platformID=char(OpenSimInstallInfo{1,1}{end,1});
+        OpenSimIs64bit = not(isempty(strfind(platformID, '64')));
+        MatlabIs64bit = not(isempty(strfind(mexext, '64')));
+        if MatlabIs64bit && not(OpenSimIs64bit)
+            h = msgbox('Matlab is 64 bit. OpenSim is 32 bit. Re-build OpenSim with 64 bit compiler.','Error');
+            uiwait(h)
+            error('Program exit')
+        elseif not(MatlabIs64bit) && OpenSimIs64bit
+            h = msgbox('Matlab is 32 bit. OpenSim is 64 bit. Re-build OpenSim with 32 bit compiler.','Error');
```
Comment (chrisdembia) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: Users of this script are not necessarily building from source. They may (in the future) obtain OpenSim binaries through chocolatey, etc.

File: Bindings/Java/Matlab/configureOpenSim.m.in [hunk 5]
```diff
@@ -49,18 +46,36 @@
...
     OpenSimLibPath  = fullfile(openSimFolder, '@OPENSIM_INSTALL_SHAREDLIBDIR@');
-% Edit the library path txt file
-    edit_path_txt_file(libraryFile,OpenSimLibPath)
+    % Edit the library path txt file
+    success = edit_path_txt_file(libraryFile, OpenSimLibPath);
+    if ~success
+        libraryFile = fullfile(prefdir, 'javalibrarypath.txt');
+        success = edit_path_txt_file(libraryFile, OpenSimLibPath);
+        if ~success
+            h = msgbox('Failed to add librarypath.', 'Error');
```
Comment (chrisdembia) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: to add *to*, in my opinion

File: Bindings/Java/Matlab/configureOpenSim.m.in [hunk 6]
```diff
@@ -49,18 +46,36 @@
...
+            uiwait(h)
+            error('Program exit');
+        end
+    end
 %% Edit Matlab path
     cleanMatlabSearchPath
     addpath(OpenSimLibPath)
```
Comment (chrisdembia) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: I'm pretty sure this line is not necessary. At least it's not necessary for me on OSX; I don't know about windows. ANyway, we can leave it in for now and delve deeper after the workshop.

**Comment by chrisdembia** [comment]
@klshrinidhi looks pretty good. I made a few comments. My biggest comment is that the former method is very unlikely to fail, and should even work in newer versions that support placing the files in `prefdir`. The thing is that the former method writes to files that we really shouldn't be editing (requires administator privileges / sudo). So I'd prefer to switch the logic, or detect the MATLAB version.

By inspecting historical matlab documentation, it is clear that the ability to set javaclasspath.txt and javalibrarypath.txt in the prefdir was introduced in 2012b

See 2012a docs which do not mention prefdir: http://www.mathworks.com/help/releases/R2012a/techdoc/matlab_external/f4863.html#f111424

And 2012b docs which do mention prefdir: http://www.mathworks.com/help/releases/R2012b/matlab/matlab_external/bringing-java-classes-and-methods-into-matlab-workspace.html?searchHighlight=javaclasspath

You can branch on matlab version with `verlessthan`: http://www.mathworks.com/help/matlab/ref/verlessthan.html

**Comment by klshrinidhi** [comment]
I think it is okay for now to leave the logic as is. Future PRs can introduce checking for Matlab version.

File: Bindings/Java/Matlab/configureOpenSim.m.in (hunk 6 above)
Comment (klshrinidhi) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: It is good leave existing logic as is for this PR.

File: Bindings/Java/Matlab/configureOpenSim.m.in (hunk 4 above)
Comment (klshrinidhi) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: This PR does not intend address those issues for now. We can worry about those later.

File: Bindings/Java/Matlab/configureOpenSim.m.in (hunk 2 above)
Comment (klshrinidhi) [inline Bindings/Java/Matlab/configureOpenSim.m.in]: Done.

File: CMakeLists.txt (hunk 1 above)
Comment (klshrinidhi) [inline CMakeLists.txt]: * Added the other part -- 32 bit Matlab && 64 bit OpenSim.
* Changed warning messages.

**Comment by chrisdembia** [comment]
Let me know if this is ready for review again.

Have you been able to test this? If so, on which OS?

**Comment by klshrinidhi** [comment]
Following:
```
import org.opensim.modeling.*
model = Model()
```
fails to load the shared library.
I verified that changes to 'classpath' and 'librarypath' are in effect. I am not sure why it fails to find the shared library there.

**Comment by chrisdembia** [comment]
Could be that `LD_LIBRARY_PATH` needs to be set also. org-opensim-modeling.jar (classpath) needs osimJavaJNI.so (librarypath), and osimJavaJNI.so needs the rest of `osim*.so` (LD_LIBRARY_PATH on Linux, PATH on Windows, not necessary on OSX b/c RPATH).

**Comment by klshrinidhi** [comment]
I got it to work:
```matlab
import org.opensim.modeling.*
model = Model()
```
Adding path to `install/lib` directory to java library path of Matlab doesn't seem to work. 
Following is working for me:
* Add paths to shared libraries to `LD_LIBRARY_PATH` or its equivalent.
* Move/rename `libstdc++` that comes with Matlab. This forces Matlab to use the library available in system standard dirs.

**Comment by chrisdembia** [comment]
@camilletown do you think you can test this script (lower priority than the hopper example)? The steps would be to build OpenSim with this branch, then open MATLAB and run the configureOpenSim script. Then, make sure you can run those two lines that @klshrinidhi wrote above. On OSX, you should not need to edit LD_LIBRARY_PATH, etc.

@klshrinidhi I will test this on Windows.

**Comment by chrisdembia** [comment]
Oops... realized I don't have MATLAB on my Windows VM. @carmichaelong would you be able to test this on Windows?

**Comment by klshrinidhi** [comment]
I will test try testing it on Windows.

**Comment by chrisdembia** [comment]
I've only looked at a little bit of the example so far. So far, I like it a lot. 

Given that there are multiple "answers" files, I think it would be better if they were in their own subdirectory.

**Comment by tkuchida** [comment]
> Given that there are multiple "answers" files, I think it would be better if they were in their own subdirectory.

I think this comment was intended for #1031. I'll move it there.

**Comment by chrisdembia** [comment]
hah oops.

**Comment by klshrinidhi** [comment]
I am able to successfully run the same commands in Matlab on Windows too.
So this PR has been tested on Linux & Windows.

**Comment by chrisdembia** [comment]
Cool. I'm merging then. If others find issues with the script, we can have additional PRs to fix them.

Did you have to set PATH on Windows?

**Comment by klshrinidhi** [comment]
Yes. PATH needs to be set on Windows.

**Pull request merged**